File: problems/0155-min-stack/solution.cpp
class MinStack {
private:
    std::stack<int> stack;      // Main stack for values.
    std::stack<int> min_stack;  // Stack to track minimums.
    
public:
    MinStack() {
        /**
         * Problem: 0155 - Min Stack
         * 
         * Design a stack with O(1) push, pop, top, and getMin operations.
         * 
         * Approach:
         * Use two stacks - one for values, one for minimums.
         * Min stack stores the minimum value at each state of the main stack.
         * This allows O(1) retrieval of minimum at any time.
         * 
         * Time complexity: O(1) for all operations.
         * Space complexity: O(n) for storing n elements.
         */
    }
    
    void push(int val) {
        // Push value to main stack.
        stack.push(val);
        
        // Push new minimum to min stack.
        // If min_stack is empty, val is the minimum.
        // Otherwise, minimum is the smaller of val and current min.
        if (min_stack.empty()) {
            min_stack.push(val);
        } else {
            min_stack.push(std::min(val, min_stack.top()));
        }
    }
    
    void pop() {
        // Remove top element from both stacks.
        stack.pop();
        min_stack.pop();
    }
    
    int top() {
        // Get top element of stack.
        return stack.top();
    }
    
    int getMin() {
        // Get minimum element in O(1) time.
        return min_stack.top();
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: problems/0155-min-stack/solution.py
class MinStack:
    def __init__(self):
        """
        Problem: 0155 - Min Stack
        
        Design a stack with O(1) push, pop, top, and getMin operations.
        
        Approach:
        Use two stacks - one for values, one for minimums.
        Min stack stores the minimum value at each state of the main stack.
        This allows O(1) retrieval of minimum at any time.
        
        Time complexity: O(1) for all operations.
        Space complexity: O(n) for storing n elements.
        """
        self.stack = []      # Main stack for values.
        self.min_stack = []  # Stack to track minimums.

    def push(self, val: int) -> None:
        """Push value onto stack and update minimum."""
        # Push value to main stack.
        self.stack.append(val)
        
        # Push new minimum to min stack.
        # If min_stack is empty, val is the minimum.
        # Otherwise, minimum is the smaller of val and current min.
        if not self.min_stack:
            self.min_stack.append(val)
        else:
            self.min_stack.append(min(val, self.min_stack[-1]))

    def pop(self) -> None:
        """Remove top element from both stacks."""
        self.stack.pop()
        self.min_stack.pop()

    def top(self) -> int:
        """Get top element of stack."""
        return self.stack[-1]

    def getMin(self) -> int:
        """Get minimum element in O(1) time."""
        return self.min_stack[-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()
